### ISSUE TITLE
feature request: page #anchoring

### ISSUE THREAD

**Issue by Pomax** [mozilla/pdf.js#5753]
Linking someone to a specific section in a 191 page document is rather cumbersome, and it would be super useful if pdf.js could dynamically update the URL so that scrolling through a document appends a #page1, #page2, etc. for linking at least to a specific page.

**Comment by Snuffleupagus** [comment]
Linking to the current position is already easy with the help of the "Current View" button, see below:

![current_view](https://cloud.githubusercontent.com/assets/2692120/6319597/005ae246-bac2-11e4-8316-c253f40699f5.png)

Closing as answered.

**Comment by Pomax** [comment]
please reopen, because that does not address the core request: #pageid suffixes in the actual URL. As handy as that bookmark button is, it generates a URL only on request. I'm requesting always having the page identifier updated in the currently visible URL (using history.replaceState for instance) so that we don't *need* to push a button just to get the link to what we're looking at. It would make for a much nicer user experience.

**Comment by timvandermeij** [comment]
I personally don't think it's a good idea to constantly update the URL (if I read your feature request correctly), but I'll reopen it for further discussion.

**Comment by Snuffleupagus** [comment]
The problem with this feature request, as I see it, is that it will not always be possible to update the hash. This would thus result in a different user experience, depending on the circumstances:
* Some browsers (mostly older ones) do not support `pushState/replaceState`.
* If you open a PDF file from the local file system in Firefox (i.e. the URL begins with `file://`), `pushState/replaceState` is *not* allowed to modify the hash. Trying to do so, results in an error: `SecurityError: The operation is insecure.`
* Finally, for very long documents, constantly updating the hash could *potentially* make the viewer sluggish when scrolling very quickly through the document.

I'm sympathetic to the feature request, but I'm not sure if it makes sense to implement this. Especially given that the functionality is, in a roundabout-way, already available.

**Comment by tycho01** [comment]
I came to this repo intending to file the same request.

I get that performance is an important consideration, although I think it wouldn't need to update for every inch scrolled to be useful -- just page/section level would be do I think.

Even if added value would be limited to files on the web / not IE, I think this simple addition would help make content on the web more 'shareable', as many casual users may not have found out about the readily available 'current view' button.

**Comment by Pomax** [comment]
Also note that enough time has passed to render the points in https://github.com/mozilla/pdf.js/issues/5753#issuecomment-75451859 effectively moot. Everything supports the history API at this point (and I am not counting browsers that reached EOL and have a valid replacement offering by the people that made that browser, similar to how we assume people are on the latest Firefox, and not FF39 or something anymore), and the performance issue during scroll has been addressed with passive event handling (although even without that I very much doubt perf would be impacted, as you don't set the hash "immediately", you schedule it to happen in the immediate future so that fast updates to what that hash should be just update the var for it, with a replaceState happening after scrolling has stopped).

2 I don't think was actually an argument against, since all that happens on file:/// is "your escalator is now a staircase". Instead of the improved hash-in-url, you just get pdf.js functionality that does exactly what it already does right now. (I'm also not sure "sharing the URL" for a local file would make a lot of sense, since you'll have to expose it over a network if you want someone else to also view it, at which point you're back to a hosted pdf where arguments around file:/// restrictions don't apply)

**Comment by tycho01** [comment]
quick PoC:
```js
let interval = 1000; // check every second.
setInterval(() => {
  let fragment = '#page1'; // TODO: get fragment from current position
  let url = PDFViewerApplication.url + fragment; // or PDFViewerApplication.pdfLinkService.baseUrl
  history.pushState(null, null, url);
}, interval);
```
open questions: 
- is there a better way to call this than to schedule at an interval? do we have any functions that's called on document scroll?
- how do we get the new fragment for the new position in the document?

**Comment by Pomax** [comment]
For the first question: we can trigger on scroll just fine, as long as we don't "do any work" until the last scroll event, using a timeout that we clear if a new scroll event comes in that invalidates it:

```
function setupHashUpdating() {
  let tid = false;

  // This will actually do the has replacement, with DOM access, pdf position querying, etc.
  let update = hash => {
    // ...heavy lifting goes here...
  };

  // We keep our test fast and lean. It's only role is to schedule the real
  // update call for some time in the future.
  let test = evt => {
    if (tid) { clearTimeout(tid); }
    tid = setTimeout(update, HASH_UPDATE_DELAY);
  };

  // And we need to make sure to tell the browser that we need passive handling.
  // See https://github.com/WICG/EventListenerOptions/blob/gh-pages/explainer.md
  window.addEventListener("scroll", test, { passive: true });
});
```

So `test` fires on scroll, in a way that doesn't interfere with the actual scrolling itself, and tries to schedule the hash update for some time in the future. Any subsequent scroll event that happens before the scheduled update occurs simply erases the current scheduled update so it never happens, and then schedules a fresh one instead. No more code than that is necessary, all the _actual_ heavy lifting happens in `update`, which won't fire and thus won't hold up the thread until after the user has stopped scrolling.

For the second question, have a look at https://github.com/mozilla/pdf.js/blob/103f4616accbd29755e6d150f0a7c2e7a6bc6546/web/app.js#L1784

**Comment by tycho01** [comment]
~hm. that `getAnchorUrl` just takes the hash as an argument, which seems to get passed from `webViewerUpdateViewarea`'s `evt.location.pdfOpenParams` (originally created in [`_updateLocation(PDFViewerApplication.pdfLinkService.page)`](https://github.com/mozilla/pdf.js/blob/98b4aff291f68adfec7209f830892f865f8ebbef/web/base_viewer.js#L798-L825)) as passed by `updateviewarea` from [`PDFViewer.update`](https://github.com/mozilla/pdf.js/blob/42b7bb47511ca30e55a41c653c6bdc9d841bfb4b/web/pdf_viewer.js#L48-L86), called even in [`_scrollUpdate()`](https://github.com/mozilla/pdf.js/blob/98b4aff291f68adfec7209f830892f865f8ebbef/web/base_viewer.js#L540-L545) so potentially a good alternative to the `scroll` event!~

whoa, looks like just `this.pdfHistory.pushCurrentPosition()` should already handle all the heavy lifting. I should compare with/without the event for performance.

naive and event-based approaches:
- https://github.com/mozilla/pdf.js/compare/master...tycho01:5753-update-hash-inline
- https://github.com/mozilla/pdf.js/compare/master...tycho01:5753-update-hash-event

**Comment by Pomax** [comment]
In the event compare, you've made `test` do too much: there is no reason to access history and then bind a new function every time test gets run. Also given that we have arrow functions, no need for `bind` anymore.

```javascript
function setupHashUpdating() {
  const HASH_UPDATE_DELAY = 1000; // there's probably "a place for constants" that we should pull from
  let tid = false;

  // set up the update function
  let updateCurrentPosition = () => PDFViewerApplication.pdfHistory.pushCurrentPosition();

  // schedule the update call for some time in the future.
  let test = (evt) => {
    if (tid) {
      clearTimeout(tid);
    }
    tid = setTimeout(updateCurrentPosition, HASH_UPDATE_DELAY);
  };

  // make sure to tell the browser that we need passive handling.
  // See https://github.com/WICG/EventListenerOptions/blob/gh-pages/explainer.md
  window.addEventListener('scroll', test, { passive: true, });
}
```
Also, it's a great idea to not call callback functions "callback" - for one this isn't really a callback (in that nothing gets called back to close a loop, this is more an interrupted code path), but having the thing be named based on what it (is supposed to) do(es) makes the code much easier to read after everyone involved's forgotten about it again =)

**Comment by tycho01** [comment]
@Pomax Thanks! Opened a PR at  #10383.

**Comment by timvandermeij** [comment]
Fixed in #10423.
